businesskaniki/nairobi-north
# Authentication

## Pull Request Description

### What does this PR do?

This pull request implements authentication and authorization features in both the frontend and backend of the application. In addition, it also introduces two essential security enhancements in the backend:

1. **Authentication and Authorization:**
   - In the frontend, we've integrated user authentication and authorization to ensure secure access control throughout the application. Users can now log in and access authorized resources based on their roles.
   - The frontend components have been updated to display content based on user roles, and protected routes are in place to ensure only authenticated users can access restricted areas.

2. **Password Reset (Forgotten Password):**
   - In the backend, we've introduced a new feature for users who forget their passwords. Users can now initiate the password reset process by providing their email address.
   - A secure and user-friendly password reset flow has been implemented, allowing users to reset their password through a link sent to their email. This feature enhances user convenience and security.

3. **Email Verification on Register:**
   - Another critical security feature added to the backend is email verification upon user registration. New users are required to verify their email address before gaining full access to the application.
   - After registration, users receive an email with a verification link, which they must click to confirm their email address. This added layer of security helps ensure that user accounts are valid and accurate.

### How should this be manually tested?

To test this pull request, follow these steps:

1. **Authentication and Authorization:**
   - Verify that only authenticated users can access protected areas of the frontend.
   - Test user registration and login functionality.
   - Check the behavior of different user roles and permissions.
   
2. **Password Reset:**
   - Try the "Forgot Password" feature by initiating a password reset request.
   - Confirm that the email with a password reset link is sent.
   - Verify that you can reset the password using the link.

3. **Email Verification:**
   - Register a new user account.
   - Check your email for a verification link and verify your email address.
   - Confirm that you gain full access to the application after email verification.

- [ ] Code follows the coding style guidelines.
